resource-directory/service/subscriptions.go

```diff
@@ -449,29 +449,65 @@ func (s *Subscriptions) cancelSubscription(localSubscriptions *sync.Map, subscri
 	return s.Close(subscriptionID, nil)
 }
 
+func (s *Subscriptions) closeSubscriptions(localSubscriptions *sync.Map) {
+	subs := make([]string, 0, 32)
+	localSubscriptions.Range(func(k interface{}, _ interface{}) bool {
+		subs = append(subs, k.(string))
+		return true
+	})
+
+	for _, sub := range subs {
+		err := s.Close(sub, nil)
+		if err != nil {
+			log.Errorf("cannot close subscription('%s'): %w", sub, err)
+		}
+	}
+}
+
+func (s *Subscriptions) getSubscriptionResponse(req *pb.SubscribeToEvents, localSubscriptions *sync.Map) (*pb.Event, bool) {
```

Comment:
this is weird - it will better when we split it to 2 functions `getSubscriptionResponse` and `getErrorSubscriptionResponse` and `getSubscriptionResponse` returns (*pb.Event, error) and when err != nil then `getErrorSubscriptionResponse` and send it.

Reply:
How to propagate `SubscriptionId` from the first error case?

```
if err != nil {
			subRes.GetOperationProcessed().ErrorStatus.Code = pb.Event_OperationProcessed_ErrorStatus_ERROR
			subRes.GetOperationProcessed().ErrorStatus.Message = err.Error()
		}
		subRes.SubscriptionId = r.GetSubscriptionId()
```

I can only think of generating a custom error type and store it there.